Add MetaData remote update into Networking and Yosemite layer for Orders and Products

pmusolino opened the pull request:
Part of #13901 

## Description
This pull request introduces functionality for updating MetaData associated with orders and products, updating just the metadata without touching the product itself. This PR manage also the local storage, associating the MetaData fetched to the right models (Order or Product) in the storage layer. It includes new remotes (Networking layer), actions, store implementations (Yosemite layer), and testing for ensuring the correct behavior of the metadata operations.

### Summary of changes
- **MetaDataAction and MetaDataStore classes added:**
  - Introduces `MetaDataAction` enum to handle actions related to updating order and product metadata.
  - Implements `MetaDataStore` to manage the storage and retrieval of metadata.

- **MetaDataRemote functionality updated:**
  - Adds `MetaDataRemote` to update just the metadata associated with an Order or a Product.
  - Adds a new file `MetaDataRemoteTests.swift` with unit tests for updating metadata for products and orders.
  - Creates a new sample JSON file `meta-data-products-orders-update.json` for testing purposes.
  - Implemented `MetaDataMapper` to parse response from `MetaDataRemote`.

## Next Steps
Given the size of this PR, I left the unit tests for `MetaDataStore` as part of another PR.

## Steps to reproduce
Right now this code is not used in the any section of the app, since it will be useful for the custom fields later on. Just unit tests should pass.

## Testing information
- Just unit tests should pass

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
Update MetaData and Networking.Media models for optional properties

- Change `metadataID` from `CopiableProp<Int64>` to `NullableCopiableProp<Int64>` in `Models+Copiable.generated.swift`
- Change `value` from `CopiableProp<String>` to `NullableCopiableProp<String>` in `Models+Copiable.generated.swift`
- Change `metadataID` from `Int64` to `Int64?` in `MetaData.swift`
- Change `value` from `String` to `String?` in `MetaData.swift`
- Update the initializer in `MetaData` to accept optional `metadataID` and `value` in `MetaData.swift`
- Modify decoding process to handle optional `metadataID` and `value` in `MetaData.swift`

pmusolino committed:
Update Product struct to handle optional metadata fields
- Modify `buildCustomFields` method in `Product.swift` to include checks for optional `metadataID` and `value`.
- Add tests in `ProductMapperTests.swift` for encoding metadata without `id` and without `value`.

pmusolino committed:
Update Order.swift to refine custom fields filtering
- Modify customFields filtering logic to exclude only keys prefixed with an underscore
- Remove check for empty values in customFields filtering

pmusolino committed:
Merge commit '67ff7f4768af459fafda36cbae3cfe977e1825d9' into issue/13901-update-API-or-models-for-adding-saving-deleting-metadata-fields-in-Order-and-Product

pmusolino committed:
revert: Update Copiable models to use non-nullable properties

- Change `metadataID` from `NullableCopiableProp<Int64>` to `CopiableProp<Int64>` in `Models+Copiable.generated.swift`
- Change `value` from `NullableCopiableProp<String>` to `CopiableProp<String>` in `Models+Copiable.generated.swift`
- Change `metadataID` from `Int64?` to `Int64` in `MetaData.swift`
- Change `value` from `String?` to `String` in `MetaData.swift`
- Update the initializer in `MetaData` struct to reflect non-optional properties in `MetaData.swift`
- Update decoding logic to ensure non-optional properties in `MetaData.swift`

pmusolino committed:
revert: Refactor Product custom fields handling and update tests
- Simplify the `buildCustomFields` method in `Product.swift`.
- Remove redundant tests for metadata encoding in `ProductMapperTests.swift`.

pmusolino committed:
revert: Update Order model to filter empty custom fields
- Modify `Order.swift` to filter out metadata with empty values in addition to keys prefixed with an underscore.

pmusolino committed:
Add meta-data products and orders JSON file
- Create new file: `meta-data-products-and-orders.json`
- Add meta data entries with IDs and values

pmusolino committed:
feat: Add MetaData functionality and tests
- Update Networking.xcodeproj to include MetaData files: MetaDataRemote.swift, MetaDataMapper.swift, MetaDataMapperTests.swift, and meta-data-products-and-orders.json.
- Create MetaDataMapper to convert Data into MetaData entities.
- Implement MetaDataRemote to update metadata for site items.
- Write tests for MetaDataMapper to ensure correct mapping from JSON responses.

pmusolino committed:
feat: Add documentation comments for the MetaDataRemoteProtocol interface

pmusolino committed:
Update path handling in MetaDataRemote.swift
- Modify path assignment to handle different MetaDataType cases
- Change path from a single definition to a switch statement for order and product types

pmusolino committed:
Add MetaDataRemote functionality and tests
- Update `MetaDataRemoteProtocol` to accept metadata as `[[String: Any]]` instead of `[[String: String]]` in `MetaDataRemote.swift`.
- Add new file `MetaDataRemoteTests.swift` with unit tests for updating metadata for products and orders.
- Add new file `meta-data-products-orders-update.json` with sample metadata for testing.

pmusolino committed:
feat: Add MetaDataAction and MetaDataStore classes

- Add MetaDataAction.swift to define actions for updating order and product metadata.
- Add MetaDataStore.swift to manage the storage and retrieval of metadata for orders and products.

pmusolino committed:
Merge commit '79620bfa0b207f0e90f260d630ef5a918b9af95f' into issue/13901-update-API-or-models-for-adding-saving-deleting-metadata-fields-in-Order-and-Product

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
fix: formatting in `MetaDataAction.swift` and `MetaDataStore.swift` by adjusting function parameter spacing

pmusolino commented:
Please let me know if you prefer that I split this PR into two smaller PRs.

pmusolino committed:
update: refactor function signatures for better readability

selanthiraiyan commented on Networking/Networking/Mapper/MetaDataMapper.swift:
```diff
@@ -0,0 +1,29 @@
+    ///
+    func map(response: Data) throws -> [MetaData] {
+        let decoder = JSONDecoder()
+        let envelope = try decoder.decode(MetaDataEnvelope.self, from: response)
```
Since [this](https://github.com/woocommerce/woocommerce-ios/pull/13929/files#diff-33032ebdb224cc4b33af7c7bc2270204221321aff5260e80e384c7ffb629ceceR31) is available as a REST request, should we update the `MetaDataMapper` to handle response with the "data" envelope by using the `hasDataEnvelope` method?

pmusolino replied:
👍 implemented here b1928b62145bce26461f688c6ff30279924c7bd2

selanthiraiyan commented on Networking/Networking/Remote/MetaDataRemote.swift:
```diff
@@ -0,0 +1,56 @@
+    /// Updates metadata for a specific item.
+    ///
+    /// - Parameters:
+    ///     - site: Site for which we'll update the metadata.
+    ///     - request: The request containing metadata to be updated.
+    ///     - type: The type of item (order or product) for which we'll update the metadata.
```
We should update the parameters to reflect the method. For example, I don't see `parentID` and `metadata`.

pmusolino replied:
Good catch, I updated it here 078f3beb43f517fdcbf266095c2a294ae54c47e6

selanthiraiyan commented on Networking/Networking/Remote/MetaDataRemote.swift:
```diff
@@ -0,0 +1,56 @@
+    ///
+    public func updateMetaData(for siteID: Int64, for parentID: Int64, type: MetaDataType, metadata: [[String: Any]]) async throws -> [MetaData] {
+
+        let parameters: [String: Any] = [ParameterKey.metaData: metadata, ParameterKey.fields: "meta_data"]
```
nit: Can we move "meta_data" to a constant enum? Or can we reuse the `ParameterKey.metaData`?

pmusolino replied:
7c567d566ff825f999e81bf9dc88622aa477b72c

selanthiraiyan commented on Yosemite/Yosemite/Actions/MetaDataAction.swift:
```diff
@@ -0,0 +1,31 @@
+///
+public enum MetaDataAction: Action {
+
+    /// Update order metadata in the local database.
```
nit: Sorry, from this comment, I don't clearly understand what the method is doing. I see that the implementation also calls the remote and not only updates the local database. 

The same applies to the following `updateProductMetaData`method as well.

pmusolino replied:
Now it should be more clear 👍 acc516e4ab78fb94f18b73c09df0052f5f57f6f8

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+
+// MARK: - MetaDataStore
+//
+public final class MetaDataStore: Store {
```
We have to initialize this store [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/AuthenticatedState.swift#L100).

pmusolino replied:
Good catch. Since it's not still in use, I didn't noticed it. Done here: 60e5e0599614cd9046890fa402e84d0cc1e3f2f2

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    /// Updates (OR Inserts) the specified ReadOnly OrderNote Entities into the Storage Layer.
```
```suggestion
    /// Updates (OR Inserts) the specified ReadOnly `MetaData` Entities into the Storage Layer.
```

pmusolino replied:
070d688840918f9391048025ec0ea7f0f120b0cd

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+    ///   - productID: ID of the product.
+    ///   - siteID: Site id of the product.
+    ///
+    private func saveMetaData(_ storage: StorageType, _ readOnlyMetaData: MetaData, productID: Int64, siteID: Int64) {
```
The `private` is not needed as this is already inside a private extension.

pmusolino replied:
👍 a7e0b1ffb4494fc1f30eaa5431493ece830bb79a

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+    ///   - orderID: ID of the order.
+    ///   - siteID: Site id of the order.
+    ///
+    private func saveMetaData(_ storage: StorageType, _ readOnlyMetaData: MetaData, orderID: Int64, siteID: Int64) {
```
The `private` is not needed as this is already inside a private extension.

pmusolino replied:
Save as above 👍 https://github.com/woocommerce/woocommerce-ios/commit/a7e0b1ffb4494fc1f30eaa5431493ece830bb79a

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    /// Updates (OR Inserts) the specified ReadOnly ProductNote Entities into the Storage Layer.
```
```suggestion
    /// Updates (OR Inserts) the specified ReadOnly `MetaData` Entities for given product ID into the Storage Layer.
```

pmusolino replied:
c7cde4afbcd815d3a6040445f60378e5d4fb9d9b 👍

selanthiraiyan reviewed (commented):
Thanks for working on this, Paolo!

I left a few comments about the code. 

> Given the size of this PR, I left the unit tests for MetaDataStore as part of another PR.

I think it is OK to include unit tests in the same PR.

pmusolino committed:
Merge commit 'fa7950dc41405519f108577bf3743d253d964df4' into issue/13901-update-API-or-models-for-adding-saving-deleting-metadata-fields-in-Order-and-Product

pmusolino committed:
Merge branch 'trunk' into issue/13901-update-API-or-models-for-adding-saving-deleting-metadata-fields-in-Order-and-Product

pmusolino committed:
update: method documentation in MetaDataRemote

pmusolino committed:
Change `ParameterKey.fields` value from `"meta_data"` to `ParameterKey.metaData` in `updateMetaData` method.

pmusolino committed:
fix:
- Change comment for updating order metadata to clarify it updates both locally and remotely.
- Change comment for updating product metadata to clarify it updates both locally and remotely.

pmusolino committed:
feat: Add MetaDataStore to the array of stores in AuthenticatedState class

pmusolino committed:
fix comment

pmusolino committed:
Refactor `saveMetaData` function visibility
- Change the visibility of `saveMetaData` from private to internal in `MetaDataStore.swift` for order and product IDs.

pmusolino committed:
feat: change comment to specify `MetaData` Entities instead of `ProductNote` Entities in MetaDataStore

pmusolino committed:
feat: Add metadata parsing and tests
- Update `MetaDataMapper.swift` to support mapping from a nested data response.
- Add tests in `MetaDataMapperTests.swift` for the new mapping functionality.
- Create `meta-data-products-and-orders_nested_in_data.json` for testing.
- Modify function signatures in `MetaDataStore.swift` for consistency.

pmusolino commented:
Ready for another pass @selanthiraiyan 🙏

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+// MARK: - Upsert MetaData for Orders and Products
+//
+private extension MetaDataStore {
+    /// Updates order metadata in the local database.
```
```suggestion
    /// Updates order metadata both remotely and in the local database..
```

selanthiraiyan commented on Yosemite/Yosemite/Stores/MetaDataStore.swift:
```diff
@@ -0,0 +1,209 @@
+        storageOrder.addToCustomFields(newStorageMetaData)
+    }
+
+    /// Updates product metadata in the local database.
```
```suggestion
    /// Updates product metadata both remotely and in the local database.
```

selanthiraiyan commented on Networking/Networking/Mapper/MetaDataMapper.swift:
```diff
@@ -0,0 +1,37 @@
+            return envelope.data.metadata.filter { !$0.key.hasPrefix("_") }
+        } else {
+            let envelope = try decoder.decode(MetaDataEnvelope.self, from: response)
+            return envelope.metadata.filter { !$0.key.hasPrefix("_") }
```
Can we add a comment to explain why we filter these values with `_` prefix?

pmusolino replied:
Good suggestion, done 3fba503a2358613a8031ab9322c3c9d2c5e1d7cb

selanthiraiyan commented on Networking/Networking/Mapper/MetaDataMapper.swift:
```diff
@@ -0,0 +1,37 @@
+    ///
+    func map(response: Data) throws -> [MetaData] {
+        let decoder = JSONDecoder()
+        if hasDataEnvelope(in: response) {
+            let envelope = try decoder.decode(DataEnvelope.self, from: response)
+            return envelope.data.metadata.filter { !$0.key.hasPrefix("_") }
+        } else {
+            let envelope = try decoder.decode(MetaDataEnvelope.self, from: response)
+            return envelope.metadata.filter { !$0.key.hasPrefix("_") }
+        }
```
I think we can do the following to prevent having the filter statement twice. 

```suggestion
         let metadata = try {
            if hasDataEnvelope(in: response) {
                let envelope = try decoder.decode(DataEnvelope.self, from: response)
                return envelope.data.metadata
            } else {
                let envelope = try decoder.decode(MetaDataEnvelope.self, from: response)
                return envelope.metadata
            }
        }()

        return metadata.filter { !$0.key.hasPrefix("_") }
```

pmusolino replied:
Done a5848d37fef0afcbfdd87b8aef75564e0bc72829 👍 thanks for the review Sharma!

selanthiraiyan reviewed (approved):
Thanks for the updates, Paolo!

I left a few more non-blocking nits/suggestions. 

I assume that you will add unit tests in a future PR. 

LGTM 🚀

pmusolino committed:
Update Yosemite/Yosemite/Stores/MetaDataStore.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

pmusolino committed:
Update Yosemite/Yosemite/Stores/MetaDataStore.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

pmusolino committed:
feat: add comment to specify why keys starting with `_` are filtered out

pmusolino committed:
refactor: modify `MetaDataMapper` to consolidate the mapping logic into a single block.
